test/Microsoft.AspNetCore.Mvc.FunctionalTests/JsonPatchInputFormatterTest.cs

```diff
@@ -0,0 +1,168 @@
+
+            // Assert
+            Assert.Equal(HttpStatusCode.OK, response.StatusCode);
+            var body = await response.Content.ReadAsStringAsync();
+            var product = JsonConvert.DeserializeObject<Product>(body);
+            Assert.NotNull(product);
+            Assert.NotNull(product.Reviews);
+            Assert.Equal(2, product.Reviews.Count);
+            Assert.Equal(4, product.Reviews[1].Rating);
+            Assert.Equal(4, product.Reviews[0].Rating);
+        }
+
+        [Fact]
+        public async Task MoveOperation_Works()
+        {
+            // Arrange
+            var input = "[{ 'op': 'move', 'path': 'Reviews/1/Rating', 'from': 'Reviews/0/Rating'}]".Replace("'", "\"");
+            var request = GetPatchRequest(input);
+
+            // Act
+            var response = await Client.SendAsync(request);
+
+            // Assert
+            Assert.Equal(HttpStatusCode.OK, response.StatusCode);
+            var body = await response.Content.ReadAsStringAsync();
+            var product = JsonConvert.DeserializeObject<Product>(body);
+            Assert.NotNull(product);
+            Assert.NotNull(product.Reviews);
+            Assert.Equal(2, product.Reviews.Count);
+            Assert.Equal(4, product.Reviews[1].Rating);
+            Assert.Equal(0, product.Reviews[0].Rating);
+        }
+
+        [Fact]
+        public async Task RemoveOperation_Works()
+        {
+            // Arrange
+            var input = "[{ 'op': 'remove', 'path': 'Reviews/0/Rating'}]".Replace("'", "\"");
+            var request = GetPatchRequest(input);
+
+            // Act
+            var response = await Client.SendAsync(request);
+
+            // Assert
+            Assert.Equal(HttpStatusCode.OK, response.StatusCode);
+            var body = await response.Content.ReadAsStringAsync();
+            var product = JsonConvert.DeserializeObject<Product>(body);
+            Assert.NotNull(product);
+            Assert.NotNull(product.Reviews);
+            Assert.Equal(2, product.Reviews.Count);
+            Assert.Equal(0, product.Reviews[0].Rating);
+        }
+
+        [Fact]
+        public async Task AddOperation_InvalidValueForProperty_AddsErrorToModelState()
+        {
+            // Arrange
+            var input = "[{ 'op': 'add', 'path': 'Reviews/-', 'value': { 'Rating': 'not-a-double' }}]".Replace("'", "\"");
+            var request = GetPatchRequest(input);
+
+            // Act
+            var response = await Client.SendAsync(request);
+
+            // Assert
+            Assert.Equal(HttpStatusCode.BadRequest, response.StatusCode);
+            var body = await response.Content.ReadAsStringAsync();
+            Assert.Equal(
+                "{\"Product\":[\"The value '{\r\n  \"Rating\": \"not-a-double\"\r\n}' is invalid for target location.\"]}",
+                body);
+        }
+
+        [Fact]
+        public async Task InvalidJsonPatch_RequestEnvelope_ResultsInBadRequest()
```

Comment:
I think that this name is confusing. I'd just mention here, either by improving the name or adding a comment, that the `add` operation requires a value and this doesn't provide it.